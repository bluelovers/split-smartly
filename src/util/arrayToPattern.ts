
let screenedSymbols: Set<string>;

export function arrayToPattern(arr: (RegExp | string)[])
{
	screenedSymbols ??= new Set('.{}[]^()+*?\\/$|'.split(''))

	return arr
		.map(s => {

			if (s instanceof RegExp)
			{
				return s.source;
			}

			return s.split('').map(s => screenedSymbols.has(s) ? '\\' + s : s).join('')
		})
		.join('|')
}
