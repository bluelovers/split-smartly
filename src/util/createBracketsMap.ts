import { IBrackets, IBracketsMap, IIncludeSeparatorMode, ISplitSettings } from '../types';
import { first } from '../util';

export function createBracketsMap<M extends IIncludeSeparatorMode>(settings: ISplitSettings<M>)
{
	let { brackets = [] as IBrackets, ignoreInsideQuotes } = settings
	if (brackets === true)
	{
		brackets = settings.defaultBrackets
	}
	else if (typeof brackets === 'object' && !Array.isArray(brackets))
	{
		brackets = Object.entries(brackets) as IBrackets
	}
	else if (typeof brackets === 'string')
	{
		brackets = brackets
			.split(',')
			.map(pairText =>
			{
				let pair = pairText.trim().split(' ')
				if (pair.length !== 2)
				{
					if (first(pair).length === 2)
					{
						pair = first(pair).split('')
					}
					else
					{
						throw new TypeError(`open and close parts of brackets should be separated by space symbol`)
					}
				}
				return pair as any
			}) as IBrackets
	}

	if (ignoreInsideQuotes)
	{
		brackets.unshift([`'`, , , true], [`"`, , , true])
	}

	settings.bracketsMap = brackets.reduce((map, [open, close, ...args]) =>
	{
		if (args.length === 1 && !settings.searchWithin)
		{
			args.unshift(undefined)
		}
		let [searchLevels = settings.searchWithin && 1, ignoreMode] = args
		if (typeof searchLevels === 'number')
		{
			searchLevels = [searchLevels]
		}
		map[open] = { open, ignoreMode, searchLevels, close: close || open }
		return map
	}, {} as IBracketsMap)

	return settings
}
