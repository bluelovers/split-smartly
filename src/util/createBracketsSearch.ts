import { IIncludeSeparatorMode, ISplitSettings } from '../types';

export function createBracketsSearch<M extends IIncludeSeparatorMode>(settings: ISplitSettings<M>)
{
	const patternParts = Object.entries(settings.bracketsMap)
		// @ts-ignore
		.flatMap(([, { close, open }]) => close !== open ? [open, close] : open)
		.concat(Object.keys(settings.mentions || {}))
		.filter(Boolean)

	const pattern = settings.arrayToPattern(patternParts)
	settings.bracketsSearch = settings.createRegExp(pattern)

	return settings
}
