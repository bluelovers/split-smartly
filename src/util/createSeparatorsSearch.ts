import { IIncludeSeparatorMode, ISplitSettings } from '../types';

export function createSeparatorsSearch<M extends IIncludeSeparatorMode>(settings: ISplitSettings<M>)
{
	const { separators } = settings

	if (typeof separators === 'string' || Array.isArray(separators))
	{
		const pattern = settings.arrayToPattern([separators].flat().filter(Boolean))
		settings.separatorSearch = settings.createRegExp(pattern)
	}
	else if (separators)
	{
		settings.separatorSearch = separators
		settings.ignoreCase = separators.ignoreCase
	}
	else
	{
		settings.separatorSearch = /empty/
	}

	return settings
}
