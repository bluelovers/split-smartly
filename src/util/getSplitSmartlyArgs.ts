import {
	IIncludeSeparatorMode,
	IParametersSplitSmartly,
	ISeparators,
	ISplitSettings,
	ISplitSettingsInput,
} from '../types';
import { first } from '../util';

export const getSplitSmartlyArgs = <M extends IIncludeSeparatorMode, M2 extends IIncludeSeparatorMode = M>(args: IParametersSplitSmartly<M>,
	extraSettings?: ISplitSettingsInput<M2>,
): [string, ISeparators, ISplitSettings<M2>] =>
{

	if (!args?.length)
	{
		throw new RangeError('empty arguments')
	}
	else if (args.length === 3)
	{
		if (!extraSettings) return args as any
	}
	else if (args.length === 1)
	{
		const arg = first(args)
		if (typeof arg === 'string')
		{
			// @ts-ignore
			args.push(',', {})
		}
		else if (Array.isArray(arg))
		{
			args.unshift(null)
			args.push({})
		}
		else if (typeof arg === 'object')
		{
			// @ts-ignore
			args.unshift(null, ',')
		}
	}
	else if (args.length === 2)
	{
		if (typeof args[0] === 'string' && args[1] instanceof RegExp)
		{
			args.push({})
		}
		else if (typeof args[1] === 'string' || Array.isArray(args[1]))
		{
			args.push({})
		}
		else
		{
			args.unshift(null)
		}
	}
	else if (args.length > 3)
	{
		throw new RangeError('Too much arguments passed to splitSmartly function!!!')
	}

	// @ts-ignore
	if (extraSettings) args[2] = { ...args[2], ...extraSettings }

	return args as any
}
