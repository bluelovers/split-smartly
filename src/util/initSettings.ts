import { IIncludeSeparatorMode, ISplitSettings } from '../types';
import { isEmpty } from '../util';

export function initSettings<M extends IIncludeSeparatorMode>(settings: ISplitSettings<M>)
{
	if (Array.isArray(settings.mentions) || typeof settings.mentions === 'string')
	{
		const mentionsMap =
			[settings.mentions].flat().filter(Boolean).reduce((map, keyword: string) =>
			{
				const key = settings.ignoreCase ? keyword.toUpperCase() : keyword
				map[key] = keyword
				return map
			}, {} as Record<string, string>)

		settings.mentions = !isEmpty(mentionsMap) && mentionsMap
	}

	return settings
		.createBracketsMap()
		.createBracketsSearch()
		.createSeparatorsSearch()
}
