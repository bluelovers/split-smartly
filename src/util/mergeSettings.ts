import { IIncludeSeparatorMode, ISplitSettings, ISplitSettingsInput } from '../types';

export function mergeSettings<M extends IIncludeSeparatorMode, M2 extends IIncludeSeparatorMode = M>(_this: ISplitSettingsInput<M>, settings: ISplitSettingsInput<M2>): ISplitSettings<M2>
{
	// @ts-ignore
	if (!settings) return _this

	// @ts-ignore
	settings = { ..._this, ...settings }
	if (['brackets', 'mentions'].some(prop => prop in settings))
	{
		settings.init()
	}

	// @ts-ignore
	return settings
}
