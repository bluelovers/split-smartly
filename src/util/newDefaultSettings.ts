import {
	EnumIncludeSeparatorMode,
	IBrackets,
	ISplitSettingsInput,
} from '../types';

export function newDefaultBrackets()
{
	return [['(', ')'], ['[', ']'], ['{', '}']] as const satisfies IBrackets
}

export function newDefaultSettings()
{
	return {
		brackets: [] as IBrackets,
		mentions: [] as string[],
		ignoreInsideQuotes: true as const,
		includeSeparatorMode: EnumIncludeSeparatorMode.INCLUDE_SEPARATOR_NONE as const,
		ignoreCase: true as const,
		trimResult: true as const,
		trimSeparators: false as const,
		//check: undefined,
		defaultBrackets: newDefaultBrackets() as IBrackets,
	} satisfies ISplitSettingsInput<EnumIncludeSeparatorMode>
}
