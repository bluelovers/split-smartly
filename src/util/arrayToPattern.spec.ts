import { IIncludeSeparatorMode, IParametersSplitSmartly } from '../types';
import { inspect } from 'util';
import splitSmartly from '../index';
import { arrayToPattern } from './arrayToPattern';

describe(`describe`, () =>
{
	([

		'_',

		'/',
		/p\d:/gi,
		['SELECT ', 'FROM '],

		'AND',

		'|',

	] as (string | RegExp | (string | RegExp)[])[])
		.forEach((argv) =>
		{

			test(inspect(argv), () =>
			{
				let actual = arrayToPattern([argv].flat() as string[]);

				expect(actual).toMatchSnapshot();
			});

		})
	;

})
