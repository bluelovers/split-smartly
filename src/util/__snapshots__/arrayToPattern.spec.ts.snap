// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe '/' 1`] = `"\\\\/"`;

exports[`describe '_' 1`] = `"_"`;

exports[`describe '|' 1`] = `"\\\\|"`;

exports[`describe 'AND' 1`] = `"AND"`;

exports[`describe /p\\d:/gi 1`] = `"p\\\\d:"`;

exports[`describe [ 'SELECT ', 'FROM ' ] 1`] = `"SELECT |FROM "`;
